File: backend/app/Models/SellRequest.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

/**
 * A customer's "sell your old jewellery" request. The status constants and the
 * guarded transitions below are the single source of truth — every script
 * (bidding close), controller, and Filament action performs its move through
 * transitionTo() so the audit trail is never bypassed.
 */
class SellRequest extends Model
{
    public const STATUS_PENDING_BIDS = 'pending_bids';
    public const STATUS_BIDDING = 'bidding';
    public const STATUS_VALUATION_REVIEW = 'valuation_review';
    public const STATUS_COMPLETED = 'completed';
    public const STATUS_EXPIRED = 'expired';
    public const STATUS_CANCELLED = 'cancelled';

    /**
     * Valid forward moves. No move may skip a stage or reverse direction.
     */
    public const ALLOWED_TRANSITIONS = [
        'pending_bids' => ['bidding', 'cancelled'],
        'bidding' => ['valuation_review', 'expired', 'cancelled'],
        'valuation_review' => ['completed', 'expired'],
        'completed' => [],
        'expired' => [],
        'cancelled' => [],
    ];

    public const BID_DURATION_HOURS = 3;

    public const SETTLEMENT_CREDIT_RATE = 0.9;

    public const SETTLEMENT_DEDUCTION_RATE = 0.1;

    public const WALLET_CREDIT_VALID_DAYS = 10;

    protected $fillable = [
        'user_id',
        'request_number',
        'item_type',
        'description',
        'city',
        'contact_phone',
        'image_path',
        'video_path',
        'status',
        'bids_start_at',
        'bids_end_at',
        'highest_bid_amount',
        'winning_bid_id',
        'admin_valuation',
        'deduction_amount',
        'wallet_credit',
        'wallet_transaction_id',
        'settlement_at',
        'result_selected_at',
        'cancelled_by',
        'cancel_reason',
    ];

    protected function casts(): array
    {
        return [
            'bids_start_at' => 'datetime',
            'bids_end_at' => 'datetime',
            'highest_bid_amount' => 'decimal:2',
            'admin_valuation' => 'decimal:2',
            'deduction_amount' => 'decimal:2',
            'wallet_credit' => 'decimal:2',
            'settlement_at' => 'datetime',
            'result_selected_at' => 'datetime',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function invitations(): HasMany
    {
        return $this->hasMany(SellInvitation::class);
    }

    public function bids(): HasMany
    {
        return $this->hasMany(SellBid::class);
    }

    public function activeBids()
    {
        return $this->bids()->where('status', 'active');
    }

    public function walletCredit(): BelongsTo
    {
        return $this->belongsTo(WalletTransaction::class, 'wallet_transaction_id');
    }

    public function auditLogs(): HasMany
    {
        return $this->hasMany(SellAuditLog::class);
    }

    public function getRouteKeyName(): string
    {
        return 'request_number';
    }

    public function transitionTo(string $to, ?string $event = null, ?array $metadata = null, ?Model $actor = null): bool
    {
        if (! in_array($to, self::ALLOWED_TRANSITIONS[$this->status] ?? [], true)) {
            return false;
        }

        $from = $this->status;

        // Guarded UPDATE — atomically wins only while the row is still in the
        // expected state. Two concurrent closers can never double-transition
        // or double-audit the same request.
        $moved = static::whereKey($this->id)
            ->where('status', $from)
            ->update(['status' => $to]);

        if ($moved === 0) {
            return false;
        }

        $this->audit($event ?? "status_{$from}_to_{$to}", $from, $to, $metadata, $actor);

        return true;
    }

    public function audit(string $event, ?string $fromStatus = null, ?string $toStatus = null, ?array $metadata = null, ?Model $actor = null): void
    {
        $this->auditLogs()->create([
            'actor_type' => $actor ? $actor->getMorphClass() : null,
            'actor_id' => $actor?->getKey(),
            'event' => $event,
            'from_status' => $fromStatus,
            'to_status' => $toStatus,
            'metadata' => $metadata,
        ]);
    }

    public function isBiddingOpen(): bool
    {
        return $this->status === self::STATUS_BIDDING
            && $this->bids_end_at !== null
            && now()->lessThan($this->bids_end_at);
    }

    public function isSettled(): bool
    {
        return $this->status === self::STATUS_COMPLETED
            && $this->wallet_transaction_id !== null;
    }

    /**
     * Deterministic winner: highest amount, then earliest submitted_at, then
     * lowest vendor id — never ambiguous, so closing is idempotent-safe.
     */
    public function resolvedWinner(): ?SellBid
    {
        return $this->activeBids()
            ->orderByDesc('amount')
            ->orderBy('submitted_at')
            ->orderBy('vendor_id')
            ->first();
    }

    public static function nextRequestNumber(): string
    {
        do {
            $number = 'SRF-'.now()->format('Ymd').'-'.strtoupper(\Illuminate\Support\Str::random(4));
        } while (static::where('request_number', $number)->exists());

        return $number;
    }
}

File: mobile_app/lib/models/sell_request.dart
/// A customer "sell your old jewellery" request, mirroring
/// `Api/SellRequestController::payload`.
class SellRequest {
  SellRequest({
    required this.id,
    required this.requestNumber,
    required this.itemType,
    this.description,
    this.city,
    this.contactPhone,
    required this.status,
    this.imageUrl,
    this.videoUrl,
    this.bidsStartAt,
    this.bidsEndAt,
    required this.biddingOpen,
    required this.bidCount,
    this.highestBidAmount,
    this.adminValuation,
    this.deductionAmount,
    this.walletCredit,
    this.settlementAt,
    this.resultSelectedAt,
    this.cancelledBy,
    this.cancelReason,
    this.createdAt,
    this.timeline = const [],
  });

  final int id;
  final String requestNumber;
  final String itemType;
  final String? description;
  final String? city;
  final String? contactPhone;
  final String status;
  final String? imageUrl;
  final String? videoUrl;
  final DateTime? bidsStartAt;
  final DateTime? bidsEndAt;
  final bool biddingOpen;
  final int bidCount;
  final double? highestBidAmount;
  final double? adminValuation;
  final double? deductionAmount;
  final double? walletCredit;
  final DateTime? settlementAt;
  final DateTime? resultSelectedAt;
  final String? cancelledBy;
  final String? cancelReason;
  final DateTime? createdAt;
  final List<SellTimelineEvent> timeline;

  bool get canCancel => status == 'pending_bids' || status == 'bidding';

  String get statusLabel {
    const map = {
      'pending_bids': 'WAITING FOR SELLERS',
      'bidding': 'BIDDING OPEN',
      'valuation_review': 'UNDER REVIEW',
      'completed': 'COMPLETED',
      'expired': 'EXPIRED',
      'cancelled': 'CANCELLED',
    };
    return map[status] ?? status.toUpperCase();
  }

  factory SellRequest.fromJson(Map<String, dynamic> json) {
    return SellRequest(
      id: json['id'] as int,
      requestNumber: json['request_number'] as String,
      itemType: json['item_type'] as String,
      description: json['description'] as String?,
      city: json['city'] as String?,
      contactPhone: json['contact_phone'] as String?,
      status: json['status'] as String,
      imageUrl: json['image_url'] as String?,
      videoUrl: json['video_url'] as String?,
      bidsStartAt: json['bids_start_at'] != null
          ? DateTime.tryParse(json['bids_start_at'] as String)
          : null,
      bidsEndAt: json['bids_end_at'] != null
          ? DateTime.tryParse(json['bids_end_at'] as String)
          : null,
      biddingOpen: json['bidding_open'] as bool? ?? false,
      bidCount: (json['bid_count'] as num?)?.toInt() ?? 0,
      highestBidAmount: (json['highest_bid_amount'] as num?)?.toDouble(),
      adminValuation: (json['admin_valuation'] as num?)?.toDouble(),
      deductionAmount: (json['deduction_amount'] as num?)?.toDouble(),
      walletCredit: (json['wallet_credit'] as num?)?.toDouble(),
      settlementAt: json['settlement_at'] != null
          ? DateTime.tryParse(json['settlement_at'] as String)
          : null,
      resultSelectedAt: json['result_selected_at'] != null
          ? DateTime.tryParse(json['result_selected_at'] as String)
          : null,
      cancelledBy: json['cancelled_by'] as String?,
      cancelReason: json['cancel_reason'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
      timeline: ((json['timeline'] as List<dynamic>?) ?? const [])
          .map((e) => SellTimelineEvent.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

class SellTimelineEvent {
  SellTimelineEvent({
    required this.event,
    this.fromStatus,
    this.toStatus,
    this.metadata = const {},
    this.at,
  });

  final String event;
  final String? fromStatus;
  final String? toStatus;
  final Map<String, dynamic> metadata;
  final DateTime? at;

  factory SellTimelineEvent.fromJson(Map<String, dynamic> json) {
    return SellTimelineEvent(
      event: json['event'] as String,
      fromStatus: json['from_status'] as String?,
      toStatus: json['to_status'] as String?,
      metadata: (json['metadata'] as Map<String, dynamic>?) ?? const {},
      at: json['at'] != null ? DateTime.tryParse(json['at'] as String) : null,
    );
  }
}
